crates/tx/src/types.rs

```diff
@@ -909,6 +899,61 @@ pub struct Header {
     /// In case a memo is not present in the transaction, a
     /// byte array filled with zeroes is present instead
     pub memo_hash: namada_core::hash::Hash,
+}
+
+impl TxCommitments {
+    /// Get the hash of this transaction's code
+    pub fn code_sechash(&self) -> &namada_core::hash::Hash {
+        &self.code_hash
+    }
+
+    /// Get the transaction data hash
+    pub fn data_sechash(&self) -> &namada_core::hash::Hash {
+        &self.data_hash
+    }
+
+    /// Get the hash of this transaction's memo
+    pub fn memo_sechash(&self) -> &namada_core::hash::Hash {
+        &self.memo_hash
+    }
+
+    pub fn hash<'a>(&self, hasher: &'a mut Sha256) -> &'a mut Sha256 {
+        hasher.update(self.serialize_to_vec());
+        hasher
+    }
+
+    /// Get the hash of this Commitments
+    pub fn get_hash(&self) -> namada_core::hash::Hash {
+        namada_core::hash::Hash(
+            self.hash(&mut Sha256::new()).finalize_reset().into(),
+        )
+    }
+}
+
+/// A Namada transaction header indicating where transaction subcomponents can
+/// be found
+#[derive(
+    Clone,
+    Debug,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshDeserializer,
+    BorshSchema,
+    Serialize,
+    Deserialize,
+    PartialEq,
+)]
+pub struct Header {
+    /// The chain which this transaction is being submitted to
+    pub chain_id: ChainId,
+    /// The time at which this transaction expires
+    pub expiration: Option<DateTimeUtc>,
+    /// A transaction timestamp
+    pub timestamp: DateTimeUtc,
+    /// The commitments to the transaction's sections
+    pub batch: HashSet<TxCommitments>,
```

Comment:
Can we use `indexmap::IndexSet` insread of `HashSet`?
I think we need to preserve the order to execute batched transactions in the same order of the insertion to the batch.

Reply:
I believe this `HashSet` (and probably every instance of `HashSet` now in the codebase) is indeed just an alias for `indexmap::set::IndexSet`. Check `namada_core::collections` where it is declared.

Reply:
ah, I see. thank you for telling me.
The `HashSet` is OK. This issue is caused by the `build_batch`.